llm/llama.go

```diff
@@ -218,366 +140,450 @@ func (ft llamaFileType) String() string {
-	params.n_batch = C.int(llm.NumBatch)
-	params.n_gqa = C.int(llm.NumGQA)
-	params.n_gpu_layers = C.int(llm.NumGPU)
-	params.main_gpu = C.int(llm.MainGPU)
-	params.low_vram = C.bool(llm.LowVRAM)
-	params.f16_kv = C.bool(llm.F16KV)
-	params.logits_all = C.bool(llm.LogitsAll)
-	params.vocab_only = C.bool(llm.VocabOnly)
-	params.use_mmap = C.bool(llm.UseMMap)
-	params.use_mlock = C.bool(llm.UseMLock)
-	params.embedding = C.bool(llm.EmbeddingOnly)
-	params.rope_freq_base = C.float(llm.RopeFrequencyBase)
-	params.rope_freq_scale = C.float(llm.RopeFrequencyScale)
+	port, err := getRandomPort()
+	if err != nil {
+		return nil, fmt.Errorf("load llm: %w", err)
+	}
 
-	if len(adapters) > 0 && llm.UseMMap {
-		log.Printf("must disable mmap to use lora adapters")
-		params.use_mmap = C.bool(false)
+	params := []string{
+		"--port", fmt.Sprintf("%d", port),
+		"--ctx-size", fmt.Sprintf("%d", opts.NumCtx),
+		"--gqa", fmt.Sprintf("%d", opts.NumGQA),
+		"--rope-freq-base", fmt.Sprintf("%f", opts.RopeFrequencyBase),
+		"--rope-freq-scale", fmt.Sprintf("%f", opts.RopeFrequencyScale),
+		"--batch-size", fmt.Sprintf("%d", opts.NumBatch),
+		"--embedding",
 	}
 
-	llm.params = &params
+	if len(adapters) > 0 {
+		// TODO: applying multiple adapters is not supported by the llama.cpp server yet
+		params = append(params, "--lora", adapters[0])
+		params = append(params, "--lora-base", model)
+	} else {
+		params = append(params, "--model", model)
+	}
 
-	cModel := C.CString(model)
-	defer C.free(unsafe.Pointer(cModel))
+	if opts.NumThread > 0 {
+		params = append(params, "--threads", fmt.Sprintf("%d", opts.NumThread))
+	}
 
-	llm.model = C.llama_load_model_from_file(cModel, params)
-	if llm.model == nil {
-		return nil, errors.New("failed to load model")
+	if !opts.F16KV {
+		params = append(params, "--memory-f32")
+	}
+	if opts.UseMLock {
+		params = append(params, "--mlock")
+	}
+	if !opts.UseMMap {
+		params = append(params, "--no-mmap")
+	}
+	if opts.UseNUMA {
+		params = append(params, "--numa")
 	}
 
-	llm.ctx = C.llama_new_context_with_model(llm.model, params)
-	if llm.ctx == nil {
-		return nil, errors.New("failed to create context")
+	// try to start llama.cpp with gpu acceleration, if it fails, fallback to CPU
+	cmd, err := llamaCmd("llama_cpp_gpu", llamaGPUBin, params)
+	if err != nil {
+		return nil, fmt.Errorf("llama_cpp_gpu command setup: %w", err)
+	}
+	err = cmd.Start()
+	if err != nil {
+		return nil, fmt.Errorf("error starting the external server: %w", err)
 	}
 
-	for _, adapter := range adapters {
-		cAdapter := C.CString(adapter)
-		defer C.free(unsafe.Pointer(cAdapter))
+	// monitor the command in a goroutine, if it fails log the error and start the CPU version
+	go func() {
+		err := cmd.Wait()
+		if err != nil {
+			if err.Error() == "signal: killed" {
+				// this is expected when the server is closed due to loading a new model
+				return
+			}
+			// TODO: what is the specific error when the GPU is not supported?
```

Comment:
still need to test this in the case of no metal support, I just manually caused errors to test the fallthrough